### ISSUE TITLE
Performance regression on depthwise_separable_conv

### ISSUE THREAD

**Issue by rootjalex** [halide/Halide#5721]
I have a branch that is ~30 commits behind master (specifically up to date with 813eadc) that reports a manual schedule of depthwise_separable_conv of:
```
Manually-tuned time: 0.369731ms
```
Meanwhile, my current master branch (specifically up to date with 3e034d6) reports a manual schedule of:
```
Manually-tuned time: 0.590547ms
```

(Not including autoscheduling times because the branch has autoscheduler changes).

Per @abadams opening this issue to see if others can reproduce / diagnose

**Comment by abadams** [comment]
Can't repro on linux or os x with llvm 11. What llvm version are you using?

**Comment by rootjalex** [comment]
LLVM 11 on linux and LLVM 10 on mac os, both experience the slowdown

**Comment by abadams** [comment]
I still can't repro. Can you check the two commits you mention in clean checkouts? I.e. not as part of a branch?

**Comment by rootjalex** [comment]
Clean checkouts on linux machine:
3e034d6:
```
Manually-tuned time: 0.52348ms
```

813eadc:
```
Manually-tuned time: 0.360595ms
```

Can confirm clean checkouts on the mac later if necessary

**Comment by rootjalex** [comment]
Hmm, it's not nearly as bad on the mac.
3e034d6:
```
Manually-tuned time: 0.426782ms
```

813eadc:
```
Manually-tuned time: 0.497357ms
```

**Comment by zvookin** [comment]
It is really important to provide hardware information. At least the number of cores. If there is any indication this might have to do with parallel scheduling, I would comment out the spin change, easiest way to do that is to set the spincount to instead of 40 at the top.

**Comment by rootjalex** [comment]
Ah sorry about that. Mac has 6, Linux has 24.

**Comment by rootjalex** [comment]
The main difference I'm finding in the generated assembly is the use of `halide_mutex_(unlock + yield + lock)` versus `halide_cond_wait`. Per @abadams , I set `max_spin_count=0` in src/runtime/thread_pool_common.h and am now seeing:
```
Manually-tuned time: 0.140013ms
```
The earlier commit also achieves approximately this performance if I set `HL_NUM_THREADS=12`.

**Comment by abadams** [comment]
Yeah, that's enough of a difference that we should probably just revert that part of the earlier change. Will make a PR in the morning.

**Comment by steven-johnson** [comment]
Did this ever get dealt with? Shall we close this?

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — some fraction of these are expected to be wrong.

**Reason (CLOSE_STALE):** The report compares two February 2021 commits, and the diagnosis was the thread pool spinning added in #5408 hammering the work-queue mutex. That spin mechanism was reworked in Dec 2024 so idle threads spin on an atomic counter off the main mutex rather than repeatedly reacquiring it, and the reported before/after commits are no longer a meaningful comparison point.

**Evidence:** The spin was introduced by a471e5957 (#5408) and replaced by c70496d32 "Don't spin on the main mutex while waiting for new work" (#8433, Dec 2024), which added `halide_cond_with_spinning` (src/runtime/thread_pool_common.h:24-70).

Closing: the thread pool spinning identified as the cause here was reworked in #8433 (c70496d32), which moved idle-thread spinning off the work-queue mutex. The 2021 commit-to-commit comparison is no longer reproducible; please open a fresh issue with current numbers if `apps/depthwise_separable_conv` still underperforms.
